Feature/exception handling

BI003 opened the pull request.

BI003 committed:
Изменил тип коллекций в хранилищах клиентов и сотрудников из List в Dictionary.
Добавил новые сервисы для работы с клиентами и сотрудниками банка (ClientService, EmployeeService).
В проект BankSystem.App добавил папку Exceptions

BI003 committed:
В сервисе “ClientService” реализованы методы:
● Добавления новых клиентов (предусмотрена валидация);
● Выбрасывает исключение если клиент моложе 18 лет (в папку Exceptions, добавлен собственный класс исключения);
● Выбрасывает исключение если у клиента нет паспортных данных;
● При добавлении нового клиента создаем ему
дефолтный(долларовый) лицевой счет;
● Метод добавления дополнительного лицевого счета ранее зарегистрированному клиенту (с валидацией);
● Метод редактирования ранее добавленного лицевого счета (соответствующая валидация);
● Метод получения выборки из хранилища с применением фильтра:
ФИО, номер телефона, номер паспорта, диапазон дат рождения;

BI003 committed:
Реализованы тесты для проверки функционала сервиса “ClientService”.

BI003 committed:
Сделал метод фильтрации более универсальным

BI003 committed:
Реализован сервис для сотрудников "EmployeeService", аналогичный "ClientService".
Добавлены идентичные методы.
Добавлен метод редактирования зарплаты "Salary".
Созданы тестовые методы для проверки работоспособности.

Dem-Ivan commented on BankSystem.App/Services/ClientService.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        private void AddDefaultAccount(Client client)
+        {
```
добавление должно происходить в хранилище а не в клиента. А перед добавлением следует проверить что клиент в хранилище существует. 
Пройдет пару лекций, мы заменим хранилище на базу данных, и уловка со ссылочным типом перестанет работать.

Dem-Ivan commented on BankSystem.App/Services/ClientService.cs:
```diff
@@ -0,0 +1,105 @@
+
+        public void EditAccount(int passport, string currencyCode, decimal newAmount)
+        {
+            var client = _clientStorage.GetAllClients().FirstOrDefault(c => c.Passport == passport);
```
Вы выгребаете всю коллекцию, хотя вам нужен только один. Что если хранилище лежит на диске и в нем 100 мил человек? Будете всех выгребать в память ради того чтобы достать одного?
Реализуйте в хранилище метод который возвращает 1 человека по паспорту. Там у вас есть доступ к коллекции непосредственно.
Проверьте весь код держа в уме эту мысль. Storage должен инкапсулировать (прятать) всю работу с коллекцией. Так чтобы при подмене словаря на базу данных не пришлось переписывать сервисы.

BI003 committed:
Добавил отдельный метод для проверки существования клиента в хранилище по паспорту - ClientExists.
Изменил добавление счёта клиента в хранилище, а не в самом клиенте. Добавил проверку на существование клиента перед добавлением.
Убрал зависимости от LINQ, используя циклы для поиска и проверки данных.

BI003 committed:
Реализовал в хранилище метод который возвращает 1 человека по паспорту - GetClientByPassport.
В методах EditAccount и AddAdditionalAccount изменил логику поиска нужного клиента по пасспорту, теперь вместо выгрузки всех клиентов в память, используется метод GetClientByPassport, который выгружает только нужного клиента.

BI003 committed:
Провёл аналогичные изменения для зранилища и сервиса сотрудников.

BI003 committed:
в EmployeeStorage изменил методы GetYoungestEmployee, GetOldestEmployee, GetAverageAge. Заменил использование LINQ на обычные циклы.
